compiler/rustc_mir_build/src/check_unsafety.rs

```diff
@@ -153,9 +179,29 @@ impl<'thir, 'tcx> Visitor<'thir, 'tcx> for UnsafetyVisitor<'tcx> {
                     self.requires_unsafe(expr.span, CallToUnsafeFunction);
                 }
             }
+            ExprKind::Field { lhs: Expr { ty, .. }, .. } => {
+                // assigning to union field is okay for AccessToUnionField
+                if let ty::Adt(adt_def, _) = ty.kind() {
+                    if adt_def.is_union() {
+                        if self.in_possible_lhs_union_assign {
+                            // FIXME: trigger AssignToDroppingUnionField unsafety if needed
+                        } else {
+                            self.requires_unsafe(expr.span, AccessToUnionField);
+                        }
+                    }
+                }
+            }
+            // don't have any special handling for AssignOp since it causes a read *and* write to lhs
+            ExprKind::Assign { lhs, rhs } => {
+                // assigning to a union is safe, check here so it doesn't get treated as a read later
+                self.in_possible_lhs_union_assign = true;
+                visit::walk_expr(self, lhs);
+                self.in_possible_lhs_union_assign = false;
```

Comment:
I think this will let invalid code like this compile:

```rust
#[derive(Copy, Clone)]
struct Foo { a: bool }

union Bar {
    a: Foo,
    b: u32,
}
fn baz(bar: Bar) {
    { bar.a }.a = true;
}
```

Reply:
It catches that, I [added a test for that](https://github.com/rust-lang/rust/pull/85263/commits/f420bd385fb4d938a372ad1c53660f21fe6178a2).